base/cvd/cuttlefish/host/commands/cvdalloc/cvdalloc.cpp

```diff
@@ -0,0 +1,147 @@
+ABSL_FLAG(int, allocate_event_socket, 0, "Socket");
+
+static void usage() {
+  LOG(ERROR) << "cvdalloc --id=id --allocate_event_socket=fd1 "
+                "--teardown_event_socket=fd2";
+  LOG(ERROR) << "Should only be invoked from run_cvd.";
+}
+
+namespace cuttlefish {
+
+Result<int> CvdallocMain(int argc, char *argv[]) {
+  std::vector<char *> args = absl::ParseCommandLine(argc, argv);
+
+  if (absl::GetFlag(FLAGS_id) == 0 ||
+      absl::GetFlag(FLAGS_teardown_event_socket) == 0 ||
+      absl::GetFlag(FLAGS_allocate_event_socket) == 0) {
+    usage();
+    return 1;
+  }
+
+  /*
+   * Explicit setuid calls seem to be required.
+   *
+   * This likely has something to do with invoking external commands,
+   * but it isn't clear why an explicit setuid(0) is necessary.
+   * It's possible a Linux kernel bug around permissions checking on tap
+   * devices may be the culprit, which we can't control.
+   *
+   * TODO: We should just use capabilities on Linux and rely on setuid as
+   * a fallback for other platforms.
+   *
+   * Save our current uid, so we can restore it to drop privileges later.
+   */
+  uid_t orig = getuid();
+
+  int id = absl::GetFlag(FLAGS_id);
+
+  pid_t pid = fork();
+  if (pid < 0) return -1;
+
+  if (pid > 0) {
+    int r = setuid(0);
+    if (r == -1) {
+      LOG(ERROR) << "Couldn't setuid root: " << strerror(errno);
+      return 1;
+    }
+
+    LOG(INFO) << "cvdalloc: allocating network resources";
+
+    CreateBridge("cvd-br");
+    CreateMobileIface(absl::StrFormat("cvd-mtap-%02d", id), id, kMobileIp);
+    CreateMobileIface(absl::StrFormat("cvd-wtap-%02d", id), id, kWirelessIp);
+    CreateEthernetIface(absl::StrFormat("cvd-etap-%02d", id), "cvd-br", true,
+                        true, false);
+
+    r = setuid(orig);
+    if (r == -1) {
+      LOG(WARNING) << "cvdalloc: couldn't drop privileges";
+      return 1;
+    }
+
+    auto sock = SharedFD::Dup(absl::GetFlag(FLAGS_allocate_event_socket));
+    if (!sock->IsOpen()) {
+      LOG(ERROR) << "cvdalloc: allocate socket is closed: " << sock->StrError();
+      return 1;
+    }
+    sock->Shutdown(SHUT_RDWR);
+
+    return 0;
+  }
+
+  LOG(INFO) << "cvdalloc (teardown): waiting to teardown";
+
+  auto sock = SharedFD::Dup(absl::GetFlag(FLAGS_teardown_event_socket));
+  if (!sock->IsOpen()) {
+    LOG(ERROR) << "cvdalloc (teardown): teardown socket is closed: "
+               << sock->StrError();
+    return 1;
+  }
+  int i;
+  if (sock->Read(&i, sizeof(i)) < 0) {
+    sock->Close();
+    return 1;
+  }
+
+  int r = setuid(0);
+  if (r == -1) {
+    LOG(ERROR) << "Couldn't setuid root: " << strerror(errno);
+    return 1;
+  }
+
+  LOG(INFO) << "cvdalloc (teardown): tearing down resources";
+
+  DestroyMobileIface(absl::StrFormat("cvd-mtap-%02d", id), id, kMobileIp);
+  DestroyMobileIface(absl::StrFormat("cvd-wtap-%02d", id), id, kWirelessIp);
+  DestroyEthernetIface(absl::StrFormat("cvd-etap-%02d", id), true, true, false);
+  DestroyBridge("cvd-br");
```

Comment:
This list is repeated here and in the parent's code. This should be encapsulated somehow to avoid that duplication.